sdk/iot/core/tests/cmocka/az_iot_telemetry_tests.c

```diff
@@ -0,0 +1,283 @@
+#define TEST_PARAMS "key=value&key_two=value2"
+#define TEST_MQTT_SPAN_BUFFER_SIZE 100
+
+#define TEST_BAD_SPAN \
+  { \
+    ._internal = { .ptr = (uint8_t*)0x7FFFFFFF, .length = -1, .capacity = 10  }, \
+  }
+#define TEST_EMPTY_OPTIONS \
+  { \
+    .module_id = AZ_SPAN_LITERAL_NULL, .user_agent = AZ_SPAN_LITERAL_NULL \
+  }
+#define TEST_VALID_OPTIONS_BOTH \
+  { \
+    .module_id = AZ_SPAN_LITERAL_FROM_STR(TEST_MODULE_ID), \
+    .user_agent = AZ_SPAN_LITERAL_FROM_STR(TEST_USER_AGENT) \
+  }
+#define TEST_VALID_OPTIONS_MODULE_ID \
+  { \
+    .module_id = AZ_SPAN_LITERAL_FROM_STR(TEST_MODULE_ID), .user_agent = AZ_SPAN_LITERAL_NULL \
+  }
+#define TEST_VALID_OPTIONS_USER_AGENT \
+  { \
+    .module_id = AZ_SPAN_LITERAL_NULL, .user_agent = AZ_SPAN_LITERAL_FROM_STR(TEST_USER_AGENT) \
+  }
+
+static const az_iot_hub_client_properties g_test_params
+    = { ._internal
+        = { .properties = AZ_SPAN_LITERAL_FROM_STR(TEST_PARAMS), .current_property = 0 } };
+
+static const char g_test_correct_topic_no_options_no_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_SUFFIX;
+static const char g_test_correct_topic_with_options_no_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_MODULES_MID
+                                "" TEST_MODULE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_USER_AGENT;
+static const char g_test_correct_topic_with_options_with_params[] = TEST_CORRECT_TOPIC_PREFIX
+    "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_MODULES_MID "" TEST_MODULE_ID
+    "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS "&" TEST_USER_AGENT;
+static const char g_test_correct_topic_no_options_with_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS;
+static const char g_test_correct_topic_with_options_module_id_with_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_MODULES_MID
+                                "" TEST_MODULE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS;
+static const char g_test_correct_topic_with_options_user_agent_with_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS
+                                "&" TEST_USER_AGENT;
+
+static const az_iot_hub_client g_test_valid_client_no_options
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_EMPTY_OPTIONS } };
+static const az_iot_hub_client g_test_valid_client_with_options_both
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_BOTH } };
+static const az_iot_hub_client g_test_valid_client_with_options_module_id
```

Comment:
Same with these: it's very hard to follow what the test story is: instead of defining globals, you should allocate these on the stack. 
I am of the opinion that UTs will not reuse much code but instead provide clarity as a single can be read to understand the scenario being tested.

Reply:
These remove the dependency on the client init calls. Primarily they don't exist yet so I need to allocated them here for now.